rules/common.smk

```diff
@@ -1,17 +1,45 @@
 import pandas as pd
+from peppy import Project, SAMPLE_NAME_COLNAME as PEP_SAMPLE_COL
 from snakemake.utils import validate
 
-report: "../report/workflow.rst"
+
+def peppy_rename(df):
+    """ Rename peppy's column for sample name identification to snakemake's. """
+    if df is None:
+        return None
+    if "sample" in df.columns and PEP_SAMPLE_COL in df.columns:
+        raise Exception("Multiple sample identifier columns present: {}".format(", ".join(["sample", PEP_SAMPLE_COL])))
+    return df.rename({PEP_SAMPLE_COL: "sample"}, axis=1)
+
+
+def peppy_units(df):
+    """ Add unit/subsample indices to peppy a data frame.  """
+    if "unit" in df.columns:
+        return df
+    def count_names(names):
+        def go(rem, n, curr, acc):
+            if rem == []:
+                return acc + [n]
+            h, t = rem[0], rem[1:]
+            return go(t, n + 1, curr, acc) if h == curr else go(t, 1, h, acc + [n])
+        return go(names[1:], 1, names[0], []) if names else []
+    df.insert(1, "unit", [i for n in count_names(list(df[PEP_SAMPLE_COL])) for i in range(1, n + 1)])
+    return df
+
 
 ###### Config file and sample sheets #####
-configfile: "config.yaml"
+p = Project("prjcfg.yaml")
```

Comment:
Do I get a dictionary with the config contents from the Project object? Then, I would introduce a new Snakemake directive

```
peppyconfig: "project.yaml"
```

that loads the project and makes it available as a global object `peppy`. From this, I would like to access samples and subannotation directly (getting rid of the boilerplate here).
However, I have doubts regarding the naming of the attributes. Why is the sample dataframe called `sheet` and the subannotation called `sample_subannotation`. Seems to be asymmetric. Why not calling the first `samples` and the second `sample_subannotation`?

Reply:
Yes you do; in fact that is basically what we're already doing... `p` is an [attmap](https://pypi.org/project/attmap/) to be precise... so `p.attribute` (or `p["attribute"]` should give you whatever configuration options are in your `project.yaml` file, already... you just say `p = Project('project.yaml')` as we've done here.


as far as getting rid of the boilerplate... the only thing the boilerplate is doing is converting the name "unit" to "sample_subannotation" and "sample" to "sample_name". one idea I had is to create a new class, say, `snakePEP`, that extends `peppy.Project` and includes this boilerplate. You would just import this object and use it as a PEP but it would handle those name conversions.

I've raised the naming issue here: https://github.com/pepkit/peppy/issues/280 -- we can probably accommodate.

Reply:
Or one could add keyword args to `Project` for renaming the two? Like `Project("project.yaml", sample_col="sample", sample_subannotation_col="unit")`?

Reply:
Or we just adapt the naming in this workflow. Although it feels weird to change the wildcards to sample_subannotation instead of unit. It somehow does not sound like the right name in this context.

Reply:
I think this is just because of what you're used to... or can you elaborate?

Reply:
sample_name+subsample_name sounds as good as sample+unit. It is just longer. I just have a problem with sample_name+sample_subannotation. That somehow does not feel right, because what we want to model in the filename pattern is not an annotation but a sequencing unit (or say a subsample).

Reply:
sample_subannotation was the sheet, though, so those are not parallel things. With the naming change [we decided on](https://github.com/pepkit/peppy/issues/280) We would end up with:

- `Project.samples` is the list of `Sample` objects
- `Sample.subsamples` is the list of `Subsample` objects
- `Project.sample_table` is the pandas data.table of samples. Samples would be indexed with the column name, `sample_name`. (previously `sample`)
- `Project.subsample_table` is the pandas data.table of subsamples. Subsamples would be indexed with the column name `subsample_name` (previously `unit`). This table also has the `sample_name` column to tie them to their `Sample` object.

So,

- `Sample.sample_name` is the name (identifier) of the sample
- `Subsample.subsample_name` is the identifier of the subsample

Reply:
Would it make sense to always index the subsample_table with (sample_name, subsample_name)? Because otherwise you cannot be sure to get a unique index. People will put things like `rep1`, `rep2` in the subsample_name column.

Reply:
In your example here, I cannot find subsample_name in the subannotation table at all.

Reply:
Another thing. Shouldn't the project.yaml read like this for consistency:
```
metadata:
    sample_table: path/to/samples.tsv
    subsample_table: path/to/subsamples.tsv
```

Reply:
> In your example here, I cannot find subsample_name in the subannotation table at all.

True -- it's optional. if you don't provide it they will be indexed numerically, which is I believe what you were doing... but you *can* include it and then pull out subsamples by name (with `get_subsample`), or index them in the table with the `subsample_name` column.

Reply:
But subsample_name alone is not necessarily unique, right? In many experimental setups, it will only be unique in combination with sample_name (e.g., when encoding the lane or replicate in subsample_name).

Reply:
`get_subsample` only exists on a `Sample` object rather than on an entire `Project`, and the subsample naming should be unique within local scope/context of a single sample (i.e., units are uniquely identified by name alone when considering just one sample.) In the table, though, the unique identification would definitely be problematic unless combined with sample like you say. Here's [an example](https://github.com/pepkit/example_peps/blob/master/example_subannotation1/sample_subannotation.csv)

Reply:
Yes, exactly. So, I would suggest to always do `set_index(("sample_name", "subsample_name"), drop=False)` on the subsample_table.